src/Security/Authorization/Policy/src/AuthorizationMiddleware.cs

```diff
@@ -57,7 +57,12 @@ public async Task Invoke(HttpContext context)
 
         // IMPORTANT: Changes to authorization logic should be mirrored in MVC's AuthorizeFilter
         var authorizeData = endpoint?.Metadata.GetOrderedMetadata<IAuthorizeData>() ?? Array.Empty<IAuthorizeData>();
-        var policy = await AuthorizationPolicy.CombineAsync(_policyProvider, authorizeData);
+
+        var policies = endpoint?.Metadata.GetOrderedMetadata<AuthorizationPolicy>() ?? Array.Empty<AuthorizationPolicy>();
+        var reqirements = endpoint?.Metadata.GetOrderedMetadata<IAuthorizationRequirement>() ?? Array.Empty<IAuthorizationRequirement>();
+
+        var policy = await AuthorizationPolicy.CombineAsync(_policyProvider, authorizeData, policies, reqirements);
```

Comment:
Could we cache the policy it on `Endpoint` as an internal property (assuming `_policyProvider` is a default one)? This seems like to quite a bit to have to do on a per-request basis.

Reply:
https://github.com/dotnet/aspnetcore/pull/39892#issuecomment-1025973214

Reply:
Unrelated to this PR but if you figure out a way to store objects per endpoint, this PR (https://github.com/dotnet/aspnetcore/pull/21675) could probably be revived too.